File: PyLife/grid.py
from grid_point import GridPoint


class Grid:
    def __init__(self, scr_w: int, scr_h: int, grid_w: int, grid_h: int):
        self._w = grid_w
        self._h = grid_h
        self._cell_w = scr_w // self._w
        self._cell_h = scr_h // self._h

    def compute_pixel(self, x: int, y: int) -> list:
        px_x = x * self._cell_w
        px_y = y * self._cell_h

        return [px_x, px_y]

    def compute_pixels(self, points: list) -> list:
        pixels = []

        for pt in points:
            px_x = pt.x * self._cell_w
            px_y = pt.y * self._cell_h
            pixels.append([px_x, px_y])

        return pixels

    def is_in_bounds(self, point: GridPoint) -> bool:
        x = point.x
        y = point.y

        return (0 <= x < self._w) and (0 <= y < self._h)

    def wrap_point(self, point: GridPoint) -> GridPoint:
        point.x %= self._w
        point.y %= self._h

        return point

    @property
    def width(self) -> int:
        return self._w

    @property
    def height(self) -> int:
        return self._h

    @property
    def cell_width(self) -> int:
        return self._cell_w

    @property
    def cell_height(self) -> int:
        return self._cell_h

File: JavaScript/JSLife/grid.js
class Grid {
  constructor(width, height, cellWidth, cellHeight) {
    this._w = width;
    this._h = height;
    this._cW = cellWidth;
    this._cH = cellHeight;
  }
  
  scalePoint(gridX, gridY) {
    let pxX = gridX * this._cW;
    let pxY = gridY * this._cH;
    
    return [pxX, pxY];
  }
  
  isInBounds(gridPoint) {
    let x = gridPoint[0];
    let y = gridPoint[1];
    
    return (0 <= x && x < this._w) && (0 <= y && y < this._h);
  }
  
  wrapPoint(gridPoint) {
    let x = gridPoint[0] % this._w;
    let y = gridPoint[1] % this._h;
    
    if (x < 0) {
      x = this._w + x;
    }
    if (y < 0) {
      y = this._h + y;
    }
    
    return [x, y];
  }
  
  wrapEntityPosition(entity) {
    entity.x %= this._w;
    entity.y %= this._h;
    if (entity.x < 0) {
      entity.x = this._w + entity.x;
    }
    if (entity.y < 0) {
      entity.y = this._h + entity.y;
    }
  }
  
  get width() {
    return this._w;
  }
  
  get height() {
    return this._h;
  }
  
  get cellWidth() {
    return this._cW;
  }
  
  get cellHeight() {
    return this._cH;
  }
}
